src/Compilers/CSharp/Portable/FlowAnalysis/ControlFlowPass.cs

```diff
@@ -310,6 +313,34 @@ public override BoundNode VisitLabeledStatement(BoundLabeledStatement node)
         public override BoundNode VisitGotoStatement(BoundGotoStatement node)
         {
             _labelsUsed.Add(node.Label);
+
+            // check for illegal jumps across using declarations
+            var sourceStart = node.Syntax.Location.SourceSpan.Start;
+            var targetStart = node.Label.Locations[0].SourceSpan.Start;
+
+            foreach (var usingDecl in _usingDeclarations)
```

Comment:
The approach is O(n^m) with n `goto`s and m `using` declarations. Would it make sense to store the using declarations in an `ArrayBuilder` in order, and then search for the `goto` and target locations using a binary search of that list?

Reply:
I started prototyping this, but it got super complicated. We need for the using declaration to be in order, but we add them all as we visit the block. That means we'd need to insert and remove sub-block entries as we descend, and I think that actually ends up being more work. We also can't just consider the ones in the current block for forward, because we might be jumping out into a parent block, but past usings we haven't visited yet.

If you're still concerned we could create a bug and come back and think about the perf later on?